consensus/src/common.rs

```diff
@@ -75,7 +96,271 @@ impl Chain<ProtocolState> for ProtocolStateChain {
         })
     }
 
+    // FIXME: this currently returns a blake2b hash. It should
+    // return a poseidon hash according to: https://github.com/MinaProtocol/mina/blob/32f529d8d8d712a44ee75be66061ce08cbdc8924/docs/specs/consensus/README.md#517-hashstate
     fn state_hash(&self) -> Option<StateHash> {
         self.top().map(|s| s.hash())
     }
 }
+
+trait Consensus {
+    type Chain;
+    fn select_secure_chain<'a>(
+        &'a self,
+        candidates: &'a [Self::Chain],
+        constants: &ConsensusConstants,
+    ) -> Result<&'a ProtocolStateChain, ConsensusError>;
+
+    fn select_longer_chain<'a>(
+        &'a self,
+        candidate: &'a ProtocolStateChain,
+    ) -> Result<&'a ProtocolStateChain, ConsensusError>;
+}
+
+// TODO: replace from checkpoint.rs
+fn is_short_range(_cs: &ProtocolStateChain) -> bool {
+    false
+}
+
+impl Consensus for ProtocolStateChain {
+    type Chain = ProtocolStateChain;
+    fn select_secure_chain<'a>(
+        &'a self,
+        candidates: &'a [Self::Chain],
+        constants: &ConsensusConstants,
+    ) -> Result<&'a ProtocolStateChain, ConsensusError> {
+        let mut tip = self;
+
+        for c in candidates {
+            if is_short_range(c) {
+                // short-range fork, select longer chain
+                tip = self.select_longer_chain(c)?;
+            } else {
+                // long-range fork, compare relative minimum window densities
+                let tip_state = self
+                    .consensus_state()
+                    .ok_or(ConsensusError::ConsensusStateNotFound)?;
+                let candidate_state = c
+                    .consensus_state()
+                    .ok_or(ConsensusError::ConsensusStateNotFound)?;
+                let tip_density =
+                    relative_min_window_density(tip_state, candidate_state, constants)?;
+                let candidate_density =
+                    relative_min_window_density(candidate_state, tip_state, constants)?;
+                match candidate_density.cmp(&tip_density) {
+                    std::cmp::Ordering::Greater => tip = c,
+                    std::cmp::Ordering::Equal => tip = self.select_longer_chain(c)?,
+                    _ => {}
+                }
+            }
+        }
+
+        Ok(tip)
```

Comment:
Use fold instead. More rusty and allows compiler to optimize better!
Tested and passing

```suggestion
        candidates.iter().fold(
            Ok(self),
            |tip, c| -> Result<&'a ProtocolStateChain, ConsensusError> {
                if is_short_range(c) {
                    // short-range fork, select longer chain
                    self.select_longer_chain(c)
                } else {
                    // long-range fork, compare relative minimum window densities
                    let tip_state = self
                        .consensus_state()
                        .ok_or(ConsensusError::ConsensusStateNotFound)?;
                    let candidate_state = c
                        .consensus_state()
                        .ok_or(ConsensusError::ConsensusStateNotFound)?;
                    let tip_density =
                        relative_min_window_density(tip_state, candidate_state, constants)?;
                    let candidate_density =
                        relative_min_window_density(candidate_state, tip_state, constants)?;
                    match candidate_density.cmp(&tip_density) {
                        std::cmp::Ordering::Greater => Ok(c),
                        std::cmp::Ordering::Equal => self.select_longer_chain(c),
                        _ => tip, // no change
                    }
                }
            },
        )
```

Reply:
This is neat, applied. Thanks